### PR TITLE
Add the check for expression evaluation in some executors

### LINKED ISSUE #16337: Lack of the check when we use expression evaluation in some executors
### Please answer these questions before submitting your issue. Thanks!
In some [executors](https://github.com/pingcap/tidb/blob/master/expression/expression.go#L327), it will use the [`VecEval`](https://github.com/pingcap/tidb/blob/master/expression/expression.go#L468) when they want to evaluate expressions. But we need to do some checks to chose whether it can use the `VeclEval` or `Eval`.

### What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
latest master [e90aac2](https://github.com/pingcap/tidb/commit/e90aac27105ba38b2d4494ce88982eea254bd6b2)

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve?

Issue Number: close #16337 

Problem Summary:
Lack of the check when we use expression evaluation in some executors.

### What is changed and how it works?
Add the check to the `EvalExpr` to choose whether it uses the `VecExpr` or `Expr`.

### Related changes
Rename `VecEval` to `EvalExpr`.
Add the check to the `EvalExpr` to choose whether it uses the `VecExpr` or `Expr`.

### Check List 

Tests 

- Unit test
- Integration test

### CODE REVIEW

File: expression/expression_test.go [hunk 1]
```diff
@@ -186,3 +188,37 @@ func tableInfoToSchemaForTest(tableInfo *model.TableInfo) *Schema {
...
+		fillColumnWithGener(eTypes[i], input, 0, nil)
+		colBuf := chunk.NewColumn(ft, 1025)
+		colBuf2 := chunk.NewColumn(ft, 1025)
+		var err error
+		ctx.GetSessionVars().EnableVectorizedExpression = false
+		err = EvalExpr(ctx, colExpr, input, colBuf)
+		if err != nil {
+			c.Fatal(err)
+		}
+		ctx.GetSessionVars().EnableVectorizedExpression = true
```
Comment (qw4990) [inline expression/expression_test.go:209]: How about adding a check `assert(colExpr.Vectorized(), IsTrue)` here?

Comment (Reminiscent) [reply expression/expression_test.go:210]: Update. PTAL @qw4990

**Comment by zz-jason** [comment]
can we tell whether the expression is evaluated in the column-oriented way or the row-oriented way?

**Comment by zz-jason** [comment]
@Reminiscent do we need to cherry-pick this bugfix to release-4.0?

**Comment by Reminiscent** [comment]
> can we tell whether the expression is evaluated in the column-oriented way or the row-oriented way?

Yes, but I don’t think it’s necessary. After `TiDB v4.0`, almost every expression has already finished the vectorized expression evaluation. We can determine by looking at the environment variable `EnableVectorizedExpression`.

**Pull request merged**